### ISSUE TITLE
What makes the gan upsampler so effective

### ISSUE THREAD

**Issue by salvadog** [lucidrains/gigagan-pytorch#7]
I find the results of the gan upsampler to be quite impressive when compared to SD upscaler and real-esrgan. However, in my personal experience, training a SR model with paired lr-hr data in a supervised manner can sometimes make it challenging to generate realistic textures on the hr image, especially when compared to pure generative models. This often results in achieving real-esrgan like outcomes. What, in your opinion, is the main concept that makes the gan upsampler so effective?

**Comment by lucidrains** [comment]
@salvadog i guess we'll slowly disentangle what is going on after this is open sourced

**Comment by lucidrains** [comment]
@salvadog we can leave this issue open and maybe someone who has been following GAN literature can chime in

**Comment by salvadog** [comment]
@lucidrains Certainly, thank you! In my current work focused on image super resolution or restoration, the top priority is to preserve the faithfulness of the input data rather than solely pursuing visually pleasing outcomes. Although cGAN-based techniques are widely used, their capabilities in terms of generating high-quality images are still somewhat restricted. Considering that diffusion-based methods inevitably modify the input data to some extent, it would be highly advantageous to introduce an upsampler similar to gigagan that can mitigate this issue.

**Comment by lucidrains** [comment]
@salvadog yea, i'll get the upsampler finished by mid this month

**Comment by ARTUROSING** [comment]
"The power of GANs in super-resolution lies in their ability to navigate the complex and nuanced landscape of high-resolution images, weaving together the delicate threads of detail and structure to create a tapestry of realistic and stunning resolution."

**Comment by onion-liu** [comment]
> "The power of GANs in super-resolution lies in their ability to navigate the complex and nuanced landscape of high-resolution images, weaving together the delicate threads of detail and structure to create a tapestry of realistic and stunning resolution."

Interesting. Where does this sentence come from?

**Comment by keyu-tian** [comment]
@salvadog imo the multi-scale discrimination contributes the most, it somehow teaches the model how to do super resolution. And i feel it provides more supervison than "progressive growing", because the model would always receive gradients from all scales.

**Comment by nbardy** [comment]
<img width="356" alt="image" src="https://github.com/lucidrains/gigagan-pytorch/assets/1278972/83c358ae-6ffc-428c-b86a-460295b3031a">

<img width="388" alt="image" src="https://github.com/lucidrains/gigagan-pytorch/assets/1278972/6d2d45ce-0175-4596-a8a2-eb6dc1010b50">

<img width="699" alt="image" src="https://github.com/lucidrains/gigagan-pytorch/assets/1278972/0da68a6c-b394-4412-ae96-d27d8e2b2c45">

**Comment by FirestName** [comment]
Hello, am I not totally lost in here thinking this "multi-scale discriminator" result in the model that could perhaps be considered "scale equivariant"? 

I mean sure, in the SG3 it's actually the generator that is equivariant (rotation, translation).
However, in the limitations, discussion, & future directions section of the paper authors state:
"In this work we modified only the generator, but it seems likely that further benefits would be available
by making the discriminator equivariant as well."

**Comment by lucidrains** [comment]
> @salvadog imo the multi-scale discrimination contributes the most, it somehow teaches the model how to do super resolution. And i feel it provides more supervison than "progressive growing", because the model would always receive gradients from all scales.

yes, I agree with this. perhaps the unet architecture for generator too

**Comment by jere357** [comment]
I feel like CLIP embeddings are the difference maker when doing the SR task
